### PR TITLE
Update NetworkPolicySpec field comments for egress

### LINKED ISSUE #107008: NetworkPolicySpec field doc is outdated
### What happened?

I tried to understand exactly what NetworkPolicy objects mean, and got lost/confused.  https://groups.google.com/g/kubernetes-sig-network/c/6k1A8l5uxLY

For example, the comments on the `PolicyTypes` field does not precisely specify semantics, and the comments on the `Ingress` and `Egress` fields omit the significance of the `PolicyTypes` field.

### What did you expect to happen?

I expected to find clear answers.

### How can we reproduce it (as minimally and precisely as possible)?

Try to figure out from the comments in `types.go` exactly how the `PolicyType` field of a `NetworkPolicySpec` interacts with the `Ingress` and `Egress` fields.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

The comments have been deficient since egress filtering was introduced, many releases ago.

</details>

### Cloud provider

<details>
N/A
</details>

### OS version

<details>
N/A
</details>

### Install tools

<details>
N/A
</details>

### Container runtime (CRI) and and version (if applicable)

<details>
N/A
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
N/A
</details>

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
This PR updates the comments on the NetworkPolicySpec fields
to clearly indicate how PolicyTypes interacts with PodSelector.

#### Which issue(s) this PR fixes:

Fixes #107008

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-network-bugs

File: pkg/apis/networking/types.go [hunk 1]
```diff
@@ -48,30 +48,44 @@ const (
...
-	// Selects the pods to which this NetworkPolicy object applies. The array of
-	// ingress rules is applied to any pods selected by this field. Multiple network
-	// policies can select the same set of pods. In this case, the ingress rules for
-	// each are combined additively. This field is NOT optional and follows standard
-	// label selector semantics. An empty podSelector matches all pods in this
-	// namespace.
+	// Selects the pods to which this NetworkPolicy object applies.
+	// A network flow is bidirectional but initiated in one direction,
+	// from a source to a destination.
+	// The whole flow is allowed or not depending on whether the initiation is allowed.
```
Comment (danwinship) [inline pkg/apis/networking/types.go:54]: "initiation" is not a standard term and it's not obvious here that you're using it to mean both the sending side and the receiving side of the connection establishment process, as opposed to just the sending side.

The phrasing that people normally use is that NetworkPolicy applies to connections, not to packets

File: pkg/apis/networking/types.go [hunk 2]
```diff
@@ -48,30 +48,44 @@ const (
...
+	// A network flow with a pod on either or both ends is allowed to be initiated
```
Comment (danwinship) [inline pkg/apis/networking/types.go:55]: so you're leaving `ipBlock` policies undefined...

File: pkg/apis/networking/types.go [hunk 3]
```diff
@@ -48,30 +48,44 @@ const (
...
+	// if both (a) initiation from the source is allowed and (b) initiation to the
+	// destination is allowed.
+	//
+	// Initiation from a source is allowed if one of the following is true.
+	// - The source is not a pod.
```
Comment (danwinship) [inline pkg/apis/networking/types.go:60]: no, you're ignoring `ipBlock`

File: pkg/apis/networking/types.go [hunk 4]
```diff
@@ -48,30 +48,44 @@ const (
...
+	// - The source is a pod and there is at least one NetworkPolicy whose `podSelector`
+	//   selects the pod, `policyTypes` includes "Egress",
+	//   and `egress` includes a rule that allows the initiation.
```
Comment (danwinship) [inline pkg/apis/networking/types.go:63]:
```suggestion
	// - The source is a pod and there is at least one NetworkPolicy whose `podSelector`
	//   selects the pod, whose `policyTypes` includes "Egress",
	//   and whose `egress` includes a rule that allows the initiation.
```

File: pkg/apis/networking/types.go [hunk 5]
```diff
@@ -48,30 +48,44 @@ const (
...
+	// - The source is a pod and there is no NetworkPolicy whose `podSelector` selects
+	//   the pod and whose `policyTypes` includes "Egress".
```
Comment (danwinship) [inline pkg/apis/networking/types.go:65]: The standard way of describing NetworkPolicy is that pods are either isolated or not isolated, and NetworkPolicy only applies to isolated pods. Pods become isolated for ingress by having an ingress NetworkPolicy select them, and pods become isolated for egress by having an egress NetworkPolicy select them.

File: pkg/apis/networking/types.go [hunk 6]
```diff
@@ -48,30 +48,44 @@ const (
...
-	// List of ingress rules to be applied to the selected pods. Traffic is allowed to
-	// a pod if there are no NetworkPolicies selecting the pod
-	// (and cluster policy otherwise allows the traffic), OR if the traffic source is
-	// the pod's local node, OR if the traffic matches at least one ingress rule
-	// across all of the NetworkPolicy objects whose podSelector matches the pod. If
-	// this field is empty then this NetworkPolicy does not allow any traffic (and serves
-	// solely to ensure that the pods it selects are isolated by default)
+	// List of ingress rules to be applied to the selected pods if `policyTypes` includes "Ingress".
+	// If this field is empty then this NetworkPolicy does not allow any incoming flows
+	// (and serves solely to ensure that the pods it selects deny ingress by default).
```
Comment (danwinship) [inline pkg/apis/networking/types.go:82]:
```suggestion
	// List of ingress rules to be applied to the selected pods (assuming `policyTypes` includes "Ingress").
	// If this field is empty then this NetworkPolicy does not allow any incoming flows
	// (and serves solely to ensure that the pods it selects are isolated for ingress).
```

File: pkg/apis/networking/types.go [hunk 7]
```diff
@@ -48,30 +48,44 @@ const (
...
 
-	// List of egress rules to be applied to the selected pods. Outgoing traffic is
-	// allowed if there are no NetworkPolicies selecting the pod (and cluster policy
-	// otherwise allows the traffic), OR if the traffic matches at least one egress rule
-	// across all of the NetworkPolicy objects whose podSelector matches the pod. If
-	// this field is empty then this NetworkPolicy limits all outgoing traffic (and serves
-	// solely to ensure that the pods it selects are isolated by default).
+	// List of egress rules to be applied to the selected pods if `policyTypes` includes "Egress".
+	// If this field is empty then this NetworkPolicy allows no outgoing flows
+	// (and serves solely to ensure that the pods it selects are isolated by default).
```
Comment (danwinship) [inline pkg/apis/networking/types.go:88]:
```suggestion
	// List of egress rules to be applied to the selected pods (assuming `policyTypes` includes "Egress").
	// If this field is empty then this NetworkPolicy allows no outgoing flows
	// (and serves solely to ensure that the pods it selects are isolated for egress).
```

File: pkg/apis/networking/types.go [hunk 8]
```diff
@@ -48,30 +48,44 @@ const (
...
 	// This field is beta-level in 1.8
```
Comment (danwinship) [inline pkg/apis/networking/types.go:89]: you can remove this comment everywhere it appears...

File: pkg/apis/networking/types.go (hunk 1 above)
Comment (MikeSpreitzer) [reply pkg/apis/networking/types.go:54]: Do you think that if I simply say "connection" that the reader will understand that this applies to UDP as well?

File: pkg/apis/networking/types.go (hunk 2 above)
Comment (MikeSpreitzer) [reply pkg/apis/networking/types.go:55]: A NetworkPolicy necessarily applies to a pod, right?  It is in the characterization of the peer that IP blocks are used.  The existing language covers the case of a connection between a pod and a non-pod.

File: pkg/apis/networking/types.go [hunk 9]
```diff
@@ -48,30 +48,44 @@ const (
...
+	//   the pod and whose `policyTypes` includes "Egress".
+	//
+	// Initiation to a destination is allowed if at least one of the following is true.
+	// - The destination is not a pod.
+	// - The destination is a pod and the source is the pod's node.
+	// - The destination is a pod and there is at least one NetworkPolicy whose
+	//   `podSelector` selects the pod, `policyTypes` includes "Ingress",
+	//   and `ingress` includes a rule that allows the initiation.
+	// - The destination is a pod and there is no NetworkPolicy whose `podSelector`
+	//   selects the pod and whose `policyTypes` includes "Ingress".
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply pkg/apis/networking/types.go:74]: The comment on the NetworkPolicy type does not get into the generated API page, right?

File: pkg/apis/networking/types.go (hunk 5 above)
Comment (MikeSpreitzer) [reply pkg/apis/networking/types.go:65]: As you say, there are two directions of isolation.  A pod is _not_ simply "isolated" or not.

File: api/openapi-spec/swagger.json [hunk 10]
```diff
@@ -13110,22 +13110,22 @@
...
...nitiated if both (a) initiation from the source is allowed and (b) initiation to the destination is allowed.\n\nInitiation from a source is allowed if one of the following is true. - The source is not a pod. - The source is a pod and there is at least one NetworkPolicy whose `podSelector`\n  selects the pod, `policyTypes` includes \"Egress\",\n  and `egress` includes a rule that allows the initiation.\n- The source is a pod and there is no NetworkPolicy whose `podSelector` selects\n  the pod and whose `policyTypes` includes \"Egress\".\n\nInitiation to a destination is allowed if at least one of the following is true. - The destination is not a pod. - The destination is a pod and the source is the pod's node. - The destination is a pod and there is at least one NetworkPolicy whose\n  `podSelector` selects the pod, `policyTypes` includes \"Ingress\",\n  and `ingress` includes a rule that allows the initiation.\n- The destination is a pod and there is no NetworkPolicy whose `podSelector`\n  selects the pod and whose `policyTypes` includes \"Ingress\".\n\nThis field is NOT optional and follows standard label selector semantics. An empty podSelector matches all pods in this namespace."
```
Comment (MikeSpreitzer) [inline api/openapi-spec/swagger.json:13128]: The rendering into the string here is pretty sucky.  is there any way to write structure in the source comments that will be nicely rendered here?

File: pkg/apis/networking/types.go (hunk 5 above)
Comment (MikeSpreitzer) [reply pkg/apis/networking/types.go:65]: Also, I think that in plain English, "isolated" is an absolute thing (no exceptions) whereas "restricted" is not absolute (there are some exceptions).  What we mean here is not an absolute but rather a matter of imposing some limits.

File: pkg/apis/networking/types.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply pkg/apis/networking/types.go:74]: See https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#networkpolicy-v1-networking-k8s-io for example.

**Comment by MikeSpreitzer** [comment]
Please see #107041 instead.
